src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -2602,17 +2602,29 @@ private void PerformSchedulingActions(IEnumerable<ScheduleResponse> responses)
                         break;
 
                     case ScheduleActionType.CreateNode:
-                        var newNodes = new List<NodeInfo>();
+                        var newNodes = new NodeInfo[response.NumberOfNodesToCreate];
+                        var nodeConfig = GetNodeConfiguration();
+
+                        // Parallelize node spawning if OOP and NodeReuse is not enabled.
+                        if (response.RequiredNodeType == NodeAffinity.OutOfProc && !_buildParameters.EnableNodeReuse)
+                        {
+                            Parallel.For(0, response.NumberOfNodesToCreate, (i) =>
+                            {
+                                newNodes[i] = _nodeManager.CreateNode(nodeConfig, response.RequiredNodeType);
+                            });
+                        }
 
                         for (int i = 0; i < response.NumberOfNodesToCreate; i++)
                         {
-                            NodeInfo createdNode = _nodeManager.CreateNode(GetNodeConfiguration(), response.RequiredNodeType);
+                            if (response.RequiredNodeType != NodeAffinity.OutOfProc || _buildParameters.EnableNodeReuse)
+                            {
+                                newNodes[i] = _nodeManager.CreateNode(nodeConfig, response.RequiredNodeType);
+                            }
 
-                            if (createdNode != null)
+                            if (newNodes[i] != null)
```

Comment:
Can we move any of this into the parallel part? Make it a helper function?

Reply:
No need.  The results are just added to final data structure and is thus fast.  Trying to make this parallel safe is more refactoring and risk.